src/hardware/mouse/mouse_common.cpp

```diff
@@ -0,0 +1,184 @@
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "mouse_common.h"
+
+#include <algorithm>
+
+#include "checks.h"
+#include "pic.h"
+
+CHECK_NARROWING();
+
+// ***************************************************************************
+// Common variables
+// ***************************************************************************
+
+MouseInfo mouse_info;
+MouseShared mouse_shared;
+MouseVideo mouse_video;
+
+// ***************************************************************************
+// Common helper calculations
+// ***************************************************************************
+
+float MOUSE_GetBallisticsCoeff(const float speed)
+{
+	// This routine provides a function for mouse ballistics
+	// (cursor acceleration), to be reused by various mouse interfaces.
+	// Since this is a DOS emulator, the acceleration model is
+	// based on a historic PS/2 mouse scaling specification, desribed
+	// for exampel here:
+	// - https://wiki.osdev.org/Mouse_Input
+
+	// Input: mouse speed
+	// Output: acceleration coefficient (1.0f for speed >= 6.0f)
+
+	// NOTE: If we don't have raw mouse input, stay with flat profile;
+	// in such case the acceleration is already handled by the host OS,
+	// adding our own could lead to hard to predict (most likely
+	// undesirable) effects
+
+	constexpr float a      = 0.017153417f;
+	constexpr float b      = 0.382477002f;
+	constexpr float lowest = 0.5f;
+
+	// Normal PS/2 mouse 2:1 scaling algorithm is just a substitution:
+	// 0 => 0, 1 => 1, 2 => 1, 3 => 3, 4 => 6, 5 => 9, other x => x * 2
+	// and the same for negatives. But we want smooth cursor movement,
+	// therefore we use approximation model (least square regression,
+	// 3rd degree polynomial, on points -6, -5, ..., 0, ... , 5, 6,
+	// here scaled to give f(6.0) = 6.0). Polynomial would be:
+	//
+	// f(x) = a*(x^3) + b*(x^1) = x*(a*(x^2) + b)
+	//
+	// This C++ function provides not the full polynomial, but rather
+	// a coefficient (0.0f ... 1.0f) calculated from supplied speed,
+	// by which the relative mouse measurement should be multiplied
+
+	if (speed > -6.0f && speed < 6.0f)
+		return std::max((a * speed * speed + b), lowest);
+	else
+		return 1.0f;
+
+	// Please consider this algorithm as yet another nod to the past,
+	// one more small touch of 20th century PC computing history :)
+}
+
+uint8_t MOUSE_GetDelayFromRateHz(const uint16_t rate_hz)
+{
+	assert(rate_hz);
+	const auto tmp = std::lround(1000.0f / MOUSE_ClampRateHz(rate_hz));
+	return static_cast<uint8_t>(tmp);
```

Comment:
if rate_hz is updates/s
-> substituting: 
(1000 ms / s) / (updates/s)

-> the seconds cancel out
ms / update

Maybe rename `tmp` to `period_in_ms`, to be more descriptive of the conversion?

Reply:
Changed.